File: backend/src/main/java/be/technobel/parsemaster/service/implementation/EmailServiceImpl.java
package be.technobel.parsemaster.service.implementation;

import be.technobel.parsemaster.exception.Exceptions;
import be.technobel.parsemaster.repository.UserRepository;
import be.technobel.parsemaster.service.declaration.EmailService;
import freemarker.template.Configuration;
import org.springframework.mail.javamail.JavaMailSender;
import org.springframework.mail.javamail.JavaMailSenderImpl;
import org.springframework.mail.javamail.MimeMessageHelper;
import org.springframework.stereotype.Service;
import org.springframework.ui.freemarker.FreeMarkerTemplateUtils;
import java.util.HashMap;

@Service
public class EmailServiceImpl implements EmailService {
  private final JavaMailSender mailSender;
  private final UserRepository userRepository;
  private final Configuration freemarkerConfig;
  private final String supportEmail;

  public EmailServiceImpl(
    JavaMailSender mailSender,
    UserRepository userRepository,
    Configuration freemarkerConfig
  ) {
    this.mailSender = mailSender;
    this.userRepository = userRepository;
    this.freemarkerConfig = freemarkerConfig;
    this.supportEmail = ((JavaMailSenderImpl)mailSender).getUsername();
  }

  private void sendMail(String to, String subject, String templatePath, Object model) {
    try {
      final var msg = mailSender.createMimeMessage();
      final var template = freemarkerConfig.getTemplate(templatePath);
      final var content = FreeMarkerTemplateUtils.processTemplateIntoString(template, model);
      final var helper = new MimeMessageHelper(msg, true);
      helper.setTo(to);
      helper.setFrom(supportEmail);
      helper.setSubject(subject);
      helper.setText(content, true);
      mailSender.send(msg);
    }
    catch(Exception ex) {
      throw Exceptions.MAIL_EXCEPTION.create();
    }
  }

  @Override
  public void passwordModificationConfirmation(String toEmail) {
    final var user = userRepository
      .findByEmail(toEmail)
      .orElseThrow(Exceptions.USER_NOT_FOUND::create);
    final var model = new HashMap<String, String>();
    model.put("pseudonym", user.getPseudonym());
    model.put("support", supportEmail);
    final var subject = "ParseMaster - Confirmation de changement de mot de passe";
    sendMail(toEmail, subject, "email/change-password.ftl", model);
  }

  @Override
  public void passwordModificationRequest(String toEmail, String passwordChangeUrl) {
    final var user = userRepository
      .findByEmail(toEmail)
      .orElseThrow(Exceptions.USER_NOT_FOUND::create);
    final var model = new HashMap<String, String>();
    model.put("pseudonym", user.getPseudonym());
    model.put("support", supportEmail);
    model.put("url", passwordChangeUrl);
    final var subject = "ParseMaster - Changement de mot de passe";
    sendMail(toEmail, subject, "email/request-password.ftl", model);
  }

  @Override
  public void userCreation(String toEmail) {
    final var user = userRepository
      .findByEmail(toEmail)
      .orElseThrow(Exceptions.USER_NOT_FOUND::create);
    final var model = new HashMap<String, String>();
    model.put("pseudonym", user.getPseudonym());
    final var subject = "ParseMaster - Confirmation de création de compte";
    sendMail(toEmail, subject, "email/user-creation.ftl", model);
  }
}

File: src/main/kotlin/be/technobel/playzone/bll/services/impl/EmailServiceImpl.kt
package be.technobel.playzone.bll.services.impl

import be.technobel.playzone.bll.services.EmailService
import be.technobel.playzone.dal.repositories.UserRepository
import org.springframework.beans.factory.annotation.Value
import org.springframework.mail.MailException
import org.springframework.mail.MailPreparationException
import org.springframework.mail.javamail.JavaMailSender
import org.springframework.mail.javamail.MimeMessageHelper
import org.springframework.stereotype.Component
import org.springframework.stereotype.Service

@Component
data class MailProperties(
	@Value("\${spring.mail.username}")
	val username: String
)

@Service
class EmailServiceImpl(
	private val emailSender: JavaMailSender,
	private val userRepository: UserRepository,
	private val mailProperties: MailProperties,
) : EmailService {
	@Throws(MailException::class)
	override fun sendPasswordModificationEmail(toEmail: String, url: String) {
		val user = userRepository.findByEmail(toEmail) ?: throw MailPreparationException("")
		val msg = emailSender.createMimeMessage()
		MimeMessageHelper(msg, "utf-8").apply {
			setTo(toEmail)
			setFrom(mailProperties.username)
			setSubject("PlayZone - Changement de mot de passe")
			setText("""
				<p>${user.fullName()},</p>
	
				<p>Nous vous informons que votre demande de modification de mot de passe a bien été prise en compte.</p>
	
				<p>Pour modifier votre mot de passe, veuillez cliquer sur le lien ci-dessous :</p>
	
				<a href="$url">$url</a>
	
				<p>Si vous n'avez pas effectué cette demande de modification ou si vous avez des questions,
				veuillez nous contacter à l'adresse ${mailProperties.username} pour obtenir de l'aide.</p>
	
				<p>Cordialement,</p>
	
				<p>L'équipe support</p>
			""".trimIndent(), true)
		}
		emailSender.send(msg)
	}

	@Throws(MailException::class)
	override fun sendPasswordModificationConfirmationEmail(toEmail: String) {
		val user = userRepository.findByEmail(toEmail) ?: throw MailPreparationException("")
		val msg = emailSender.createMimeMessage()
		MimeMessageHelper(msg, "utf-8").apply {
			setTo(toEmail)
			setFrom(mailProperties.username)
			setSubject("PlayZone - Confirmation de changement de mot de passe")
			setText("""
				<p>${user.fullName()},</p>

				<p>Nous vous confirmons que votre mot de passe a été modifié avec succès.</p>
	
				<p>Si vous n'avez pas effectué cette modification ou si vous avez des questions,
				veuillez nous contacter à l'adresse ${mailProperties.username} pour obtenir de l'aide.</p>
	
				<p>Cordialement,</p>
	
				<p>L'équipe support</p>
			""".trimIndent(), true)
		}
		emailSender.send(msg)
	}
}